synapse/lib/layer.py

```diff
@@ -4093,6 +4094,103 @@ async def iterNodeDataKeys(self, buid):
             prop = self.getAbrvProp(abrv)
             yield prop[0]
 
+    async def iterLayerAddPerms(self):
+
+        # nodes & props
+        for byts, abrv in self.propabrv.slab.scanByFull(db=self.propabrv.name2abrv):
+            form, prop = s_msgpack.un(byts)
+            if form is None:
+                continue
+
+            if self.layrslab.prefexists(abrv, db=self.byprop):
+                if prop:
+                    yield ('node', 'prop', 'set', f'{form}:{prop}')
+                else:
+                    yield ('node', 'add', form)
+
+        # tagprops
+        for byts, abrv in self.tagpropabrv.slab.scanByFull(db=self.tagpropabrv.name2abrv):
+            info = s_msgpack.un(byts)
+            if None in info or len(info) != 3:
+                continue
+
+            if self.layrslab.prefexists(abrv, db=self.bytagprop):
+                yield ('node', 'tag', 'add', *info[1].split('.'))
+
+        # nodedata
+        for abrv in self.dataslab.scanKeys(db=self.dataname):
+            name, _ = self.getAbrvProp(abrv)
+            yield ('node', 'data', 'set', name)
+
+        # edges
+        for verb in self.layrslab.scanKeys(db=self.byverb):
+            yield ('node', 'edge', 'add', verb.decode())
+
+        # tags
+        # NB: tag perms should be yielded for every leaf on every node in the layer
+        tags = await s_spooled.Dict.anit(cell=self.core)
```

Comment:
I'm curious if you did any rough perf testing on using the spools (especially if we fallback to disk i/o) versus just letting duplicate perms through, especially since there is also the permcache.

Reply:
I did not do any perf testing, it just seemed like a good idea. We need this dict for collecting all tags on each node. The spooled set below is just for deduping perms though, I can remove it.